### ISSUE TITLE
JDK24 java/lang/Thread/virtual/stress/LotsOfContendedMonitorEnter.java#LM_LIGHTWEIGHT and a few other tests timed out

### ISSUE THREAD

**Issue by JasonFengJ9** [eclipse-openj9/openj9#20705]
Failure link
------------

From [an internal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/27/consoleFull)(`ubu22-aarch64-2`):
```
00:07:31  openjdk version "24-internal" 2025-03-18
00:07:31  OpenJDK Runtime Environment (build 24-internal-adhoc.jenkins.BuildJDKnextaarch64linuxPersonal)
00:07:31  Eclipse OpenJ9 VM (build master-4ee9b77dbbd, JRE 24 Linux aarch64-64-Bit Compressed References 20241128_142 (JIT enabled, AOT enabled)
00:07:31  OpenJ9   - 4ee9b77dbbd
00:07:31  OMR      - 494474f6819
00:07:31  JCL      - 91383efdb6c based on jdk-24+25)
```
[Rerun in Grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/parambuild/?SDK_RESOURCE=customized&TARGET=testList+TESTLIST=jdk_lang_1,jdk_foreign_0&BASE_DOCKER_REGISTRY_CREDENTIAL_ID=&TEST_FLAG=&UPSTREAM_TEST_JOB_NAME=Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal&DOCKER_REQUIRED=false&ACTIVE_NODE_TIMEOUT=&VENDOR_TEST_DIRS=functional%2Cjck&EXTRA_DOCKER_ARGS=&TKG_OWNER_BRANCH=adoptium%3Amaster&OPENJ9_SYSTEMTEST_OWNER_BRANCH=eclipse%3Amaster&PLATFORM=aarch64_linux&GENERATE_JOBS=true&KEEP_REPORTDIR=true&PERSONAL_BUILD=false&DOCKER_REGISTRY_DIR=&ADOPTOPENJDK_REPO=https%3A%2F%2Fgithub.com%2Fadoptium%2Faqa-tests.git&RERUN_ITERATIONS=0&SETUP_JCK_RUN=false&DOCKER_REGISTRY_URL_CREDENTIAL_ID=&LABEL=&EXTRA_OPTIONS=&BUILD_IDENTIFIER=fengj%40ca.ibm.com&CUSTOMIZED_SDK_URL=https%3A%2F%2Fna-public.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2FBuild_JDKnext_aarch64_linux_Personal%2F142%2FOpenJ9-JDKnext-aarch64_linux-20241128-015723.tar.gz+https%3A%2F%2Fna-public.artifactory.swg-devops.com%2Fartifactory%2Fsys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com%2FBuild_JDKnext_aarch64_linux_Personal%2F142%2Ftest-images.tar.gz&JENKINS_KEY=Jenkins+4096+key&ADOPTOPENJDK_BRANCH=master&LIGHT_WEIGHT_CHECKOUT=true&USE_JRE=false&ARTIFACTORY_SERVER=na-public.artifactory.swg-devops&KEEP_WORKSPACE=false&USER_CREDENTIALS_ID=83181e25-eea4-4f55-8b3e-e79615733226&JDK_VERSION=next&DOCKER_REGISTRY_URL=&ITERATIONS=1&VENDOR_TEST_REPOS=git%40github.ibm.com%3Aruntimes%2Ftest.git%2Cgit%40github.ibm.com%3Aruntimes%2Fjck.git&JDK_REPO=git%40github.com%3Aibmruntimes%2Fopenj9-openjdk-jdk.git&JCK_GIT_BRANCH=main&OPENJ9_BRANCH=master&OPENJ9_SHA=4ee9b77dbbdc1508639551e13bbffe7a455b5c10&JCK_GIT_REPO=&VENDOR_TEST_BRANCHES=master%2Cmain&OPENJ9_REPO=git%40github.com%3Aeclipse%2Fopenj9.git&UPSTREAM_JOB_NAME=&CLOUD_PROVIDER=&CUSTOM_TARGET=&VENDOR_TEST_SHAS=&JDK_BRANCH=openj9&LABEL_ADDITION=ci.project.openj9&ARTIFACTORY_REPO=sys-rt-generic-local%2Fhyc-runtimes-jenkins.swg-devops.com&ARTIFACTORY_ROOT_DIR=&UPSTREAM_TEST_JOB_NUMBER=43&DOCKERIMAGE_TAG=&JDK_IMPL=openj9&TEST_TIME=&SSH_AGENT_CREDENTIAL=83181e25-eea4-4f55-8b3e-e79615733226&AUTO_DETECT=true&SLACK_CHANNEL=%23rt-jenkins&DYNAMIC_COMPILE=false&RELATED_NODES=&ADOPTOPENJDK_SYSTEMTEST_OWNER_BRANCH=adoptium%3Amaster&APPLICATION_OPTIONS=&CUSTOMIZED_SDK_URL_CREDENTIAL_ID=7c1c2c28-650f-49e0-afd1-ca6b60479546&ARCHIVE_TEST_RESULTS=false&NUM_MACHINES=&OPENJDK_SHA=&TRSS_URL=https%3A%2F%2Ftrssrtp1.fyre.ibm.com%2F&USE_TESTENV_PROPERTIES=false&BUILD_LIST=openjdk&ADDITIONAL_ARTIFACTS_REQUIRED=&UPSTREAM_JOB_NUMBER=&STF_OWNER_BRANCH=adoptium%3Amaster&TIME_LIMIT=20&JVM_OPTIONS=&PARALLEL=None) - Change TARGET to run only the failed test targets.

Optional info
-------------

Failure output (captured from console output)
---------------------------------------------
```
00:09:01  ===============================================
00:09:01  Running test jdk_lang_1 ...
00:09:01  ===============================================
00:09:01  jdk_lang_1 Start Time: Thu Nov 28 05:09:00 2024 Epoch Time (ms): 1732770540869
00:09:01  variation: -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage Mode650
00:09:01  JVM_OPTIONS:  -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage -XX:-UseCompressedOops -Xverbosegclog 

00:45:24  TEST: java/lang/Thread/virtual/stress/LotsOfContendedMonitorEnter.java#LM_LIGHTWEIGHT
00:45:24  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 967799ms).

00:45:24  TEST: java/lang/Thread/virtual/stress/LotsOfContendedMonitorEnter.java#default
00:45:24  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 970810ms).

01:02:56  TEST: java/lang/Thread/virtual/MiscMonitorTests.java#default
01:02:56  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 975659ms).

01:03:17  TEST: java/lang/Thread/virtual/MiscMonitorTests.java#Xcomp
01:03:17  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 979768ms).

01:11:48  TEST: java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java#id0
01:11:48  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 2400000ms, elapsed time including timeout handling was 2416108ms).

01:19:48  TEST: java/lang/Thread/virtual/MiscMonitorTests.java#Xcomp-noTieredCompilation
01:19:48  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 1016084ms).

01:19:55  TEST: java/lang/Thread/virtual/MiscMonitorTests.java#Xcomp-TieredStopAtLevel3
01:19:55  TEST RESULT: Error. Program `/home/jenkins/workspace/Test_openjdknext_j9_sanity.openjdk_aarch64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java' timed out (timeout set to 960000ms, elapsed time including timeout handling was 996877ms).


01:44:37  jdk_lang_1_FAILED
```
[50x grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/45335/) - all failed

**Comment by babsingh** [comment]
@fengxue-IS Assigning it to you; this is related to the new VT pinning work.

**Comment by babsingh** [comment]
@thallium Can you look at this issue? The failure should be a timeout. You will need to have the latest changes in your repos (OpenJ9, OMR and extensions) and employ `-XX:+YieldPinnedVirtualThreads` while running the tests. Three different tests fail in this issue: `LotsOfContendedMonitorEnter`, `MiscMonitorTests` and `Skynet100kWithMonitors`.

**Comment by thallium** [comment]
I'll take a look.

**Comment by thallium** [comment]
@babsingh Looks like the current failure is a crash:
```
Unhandled exception
Type=Segmentation error vmState=0x00000000
J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
Handler1=0000785D9743FF80 Handler2=0000785D97A777C0 InaccessibleAddress=0000000000000000
RDI=0000785D90044A40 RSI=0000000000305AC8 RAX=0000000000000000 RBX=0000785D90349A8B
RCX=0000000000000000 RDX=00000000002A7040 R8=0000000000000000 R9=0000785D95FF7890
R10=00000000002F1100 R11=0000000000000293 R12=0000000000000000 R13=00000000002F1100
R14=0000000000000000 R15=0000785D8F7FF8B0
RIP=0000785D974D278F GS=0000 FS=0000 RSP=0000785D8F7FF340
EFlags=0000000000010246 CS=0033 RBP=00000000FFE9DE58 ERR=0000000000000004
TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000000
xmm0=0000000000305960 (f: 3168608.000000, d: 1.565500e-317)
xmm1=0000785d900154d0 (f: 2416006400.000000, d: 6.538625e-310)
xmm2=0000785d97618250 (f: 2539749888.000000, d: 6.538631e-310)
xmm3=0000785d96010da0 (f: 2516651520.000000, d: 6.538630e-310)
xmm4=0000785d975f5f70 (f: 2539609856.000000, d: 6.538631e-310)
xmm5=0000000000305968 (f: 3168616.000000, d: 1.565504e-317)
xmm6=0000000000305960 (f: 3168608.000000, d: 1.565500e-317)
xmm7=00000000ffe9fd00 (f: 4293524736.000000, d: 2.121283e-314)
xmm8=0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm9=0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm10=0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm11=0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm12=0000000000302929 (f: 3156265.000000, d: 1.559402e-317)
xmm13=00000000000be398 (f: 779160.000000, d: 3.849562e-318)
xmm14=0000000000000000 (f: 0.000000, d: 0.000000e+00)
xmm15=0000000000000000 (f: 0.000000, d: 0.000000e+00)
Module=/home/gengchen/openj9-openjdk-jdk24/build/linux-x86_64-server-release/jdk/lib/default/libj9vm29.so
Module_base_address=0000785D97400000
Target=2_90_20250319_000000 (Linux 6.8.0-45-generic)
CPU=amd64 (8 logical CPUs) (0x3e77f8000 RAM)
----------- Stack Backtrace -----------
_ZN32VM_BytecodeInterpreterCompressed3runEP10J9VMThread+0x1d72f (0x0000785D974D278F [libj9vm29.so+0xd278f])
bytecodeLoopCompressed+0xce (0x0000785D974B4FCE [libj9vm29.so+0xb4fce])
 (0x0000785D9757A932 [libj9vm29.so+0x17a932])
---------------------------------------
```

**Comment by thallium** [comment]
The crash happened in `VM_BytecodeInterpreterFull::buildInternalNativeStackFrame()` caused by a null `_sendMethod`.

**Comment by babsingh** [comment]
How does the full native stack look like?

**Comment by thallium** [comment]
Native stack with #21428 
```
 frame #0: 0x0000000100486708 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) [inlined] VM_BytecodeInterpreterFull::buildInternalNativeStackFrame(this=0x0000000170f467b0, _sp=<unavailable>, _pc=<unavailable>) at BytecodeInterpreter.hpp:487:18 [opt]
    frame #1: 0x00000001004866d4 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) [inlined] VM_BytecodeInterpreterFull::enterObjectMonitor(this=0x0000000170f467b0, _sp=<unavailable>, _pc=<unavailable>, obj=0x00000004ffe5a4d8) at BytecodeInterpreter.hpp:1204:6 [opt]
    frame #2: 0x00000001004866d4 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) at BytecodeInterpreter.hpp:8937:22 [opt]
    frame #3: 0x00000001004866d4 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(this=0x0000000170f467b0, vmThread=<unavailable>) at BytecodeInterpreter.hpp:11995:4 [opt]
    frame #4: 0x000000010045feb4 libj9vm29.dylib`bytecodeLoopFull(currentThread=<unavailable>) at BytecodeInterpreter.inc:112:21 [opt]
    frame #5: 0x00000001004b55c0 libj9vm29.dylib`cInterpreter + 16
```

Without #21428 
```
 frame #0: 0x000000010047ef2c libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) [inlined] VM_BytecodeInterpreterFull::buildInternalNativeStackFrame(this=0x0000000170eae7b0, _sp=<unavailable>, _pc=<unavailable>) at BytecodeInterpreter.hpp:487:18 [opt]
    frame #1: 0x000000010047eef8 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) [inlined] VM_BytecodeInterpreterFull::yieldPinnedContinuation(this=0x0000000170eae7b0, _sp=<unavailable>, _pc=<unavailable>, newThreadState=12, returnState=0) at BytecodeInterpreter.hpp:1532:4 [opt]
    frame #2: 0x000000010047eef8 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(J9VMThread*) at BytecodeInterpreter.hpp:8957:11 [opt]
    frame #3: 0x000000010047eef8 libj9vm29.dylib`VM_BytecodeInterpreterFull::run(this=0x0000000170eae7b0, vmThread=<unavailable>) at BytecodeInterpreter.hpp:11995:4 [opt]
    frame #4: 0x0000000100459104 libj9vm29.dylib`bytecodeLoopFull(currentThread=<unavailable>) at BytecodeInterpreter.inc:112:21 [opt]
    frame #5: 0x00000001004ad5c8 libj9vm29.dylib`cInterpreter + 16
```

**Comment by thallium** [comment]
Test passed with `-Xint`.

**Comment by babsingh** [comment]
Can you following the guidance in https://github.com/eclipse-openj9/openj9/issues/21408#issuecomment-2751397415 to help us understand the context in which the virtual threads are crashing?

**Comment by thallium** [comment]
For test `java/lang/Thread/virtual/MiscMonitorTests.java`, same crash as https://github.com/eclipse-openj9/openj9/issues/20705#issuecomment-2741712628 occurred.

With `-Xint`, crash occurred on GC worker thread:
```
* thread #16, name = 'GC Worker', stop reason = EXC_BAD_ACCESS (code=2, address=0x1702e7ff8)
  * frame #0: 0x00000001004d3264 libj9vm29.dylib`j9stackmap_StackBitsForPC [inlined] outputStackMap(liveStack=0x0000000170329070, newStackDescription=0x200060000219afdc, bits=18446744073709518320) at stackmap.c:574:28 [opt]
    frame #1: 0x00000001004d3208 libj9vm29.dylib`j9stackmap_StackBitsForPC(portLib=0x00000001001d1ed8, pc=<unavailable>, romClass=<unavailable>, romMethod=0x000000013802892c, resultArrayBase=0x000060000219c020, resultArraySize=18446744073709551615, userData=0x000000012b015620, getBuffer=<unavailable>, releaseBuffer=(libj9vm29.dylib`j9mapmemory_ReleaseBuffer at mapmemorybuffer.c:71)) at stackmap.c:136:8 [opt]
    frame #2: 0x000000010043aeb8 libj9vm29.dylib`walkBytecodeFrameSlots [inlined] getStackMap(walkState=0x000000017032a628, romClass=0x0000000138028510, romMethod=0x000000013802892c, offsetPC=70, pushCount=18446744073709551615, result=0x000060000219c020) at swalk.c:1555:14 [opt]
    frame #3: 0x000000010043ae84 libj9vm29.dylib`walkBytecodeFrameSlots(walkState=0x000000017032a628, method=<unavailable>, offsetPC=70, pendingBase=0x000000011d841c70, pendingStackHeight=18446744073709551615, localBase=<unavailable>, numberOfLocals=<unavailable>, alwaysLocalMap=<unavailable>) at swalk.c:932:3 [opt]
    frame #4: 0x000000010043a750 libj9vm29.dylib`walkStackFrames at swalk.c:1035:4 [opt]
    frame #5: 0x000000010043a688 libj9vm29.dylib`walkStackFrames(currentThread=0x000000012b825700, walkState=0x000000017032a628) at swalk.c:353:6 [opt]
    frame #6: 0x000000010045ec1c libj9vm29.dylib`walkContinuationStackFrames(currentThread=0x000000012b825700, continuation=0x000000011cc068e0, threadObject=0x0000000000000000, walkState=0x000000017032a628) at ContinuationHelpers.cpp:546:8 [opt]
    frame #7: 0x0000000100b27038 libj9gc29.dylib`GC_VMThreadStackSlotIterator::scanContinuationSlots(vmThread=<unavailable>, continuationObjectPtr=<unavailable>, userData=<unavailable>, oSlotIterator=<unavailable>, includeStackFrameClassReferences=<unavailable>, trackVisibleFrameDepth=<unavailable>) at VMThreadStackSlotIterator.cpp:152:2 [opt]
    frame #8: 0x0000000100b06004 libj9gc29.dylib`MM_ScavengerDelegate::scanContinuationNativeSlots(this=0x000000012a80dc80, env=0x000000012b827668, objectPtr=0x00000004ffde2a00, reason=SCAN_REASON_SCAVENGE, beingMounted=false) at ScavengerDelegate.cpp:395:3 [opt]
    frame #9: 0x0000000100b061d8 libj9gc29.dylib`MM_ScavengerDelegate::getObjectScanner(this=0x000000012a80dc80, env=0x000000012b827668, objectPtr=0x00000004ffde2a00, allocSpace=0x00000001200e8b28, flags=2, reason=SCAN_REASON_SCAVENGE, shouldRemember=0x00000001200e8c00) at ScavengerDelegate.cpp:496:21 [opt]
    frame #10: 0x0000000100ad76f4 libj9gc29.dylib`MM_Scavenger::incrementalScanCacheBySlot(MM_EnvironmentStandard*, MM_CopyScanCacheStandard*) [inlined] MM_Scavenger::getObjectScanner(this=0x000000012a80dc20, env=0x000000012b827668, objectptr=0x00000004ffde2a00, objectScannerState=0x00000001200e8b28, flags=2, reason=SCAN_REASON_SCAVENGE, shouldRemember=0x00000001200e8c00) at Scavenger.cpp:1902:19 [opt]
    frame #11: 0x0000000100ad76d8 libj9gc29.dylib`MM_Scavenger::incrementalScanCacheBySlot(MM_EnvironmentStandard*, MM_CopyScanCacheStandard*) [inlined] MM_Scavenger::incrementalScavengeObjectSlots(this=0x000000012a80dc20, env=0x000000012b827668, objectPtr=0x00000004ffde2a00, scanCache=0x00000001200e8af0) at Scavenger.cpp:2136:20 [opt]
    frame #12: 0x0000000100ad7368 libj9gc29.dylib`MM_Scavenger::incrementalScanCacheBySlot(this=0x000000012a80dc20, env=<unavailable>, scanCache=0x00000001200e8af0) at Scavenger.cpp:2491:46 [opt]
    frame #13: 0x0000000100ad7ef4 libj9gc29.dylib`MM_Scavenger::completeScan(this=0x000000012a80dc20, env=0x000000012b827668) at Scavenger.cpp:2567:4 [opt]
    frame #14: 0x0000000100ad8304 libj9gc29.dylib`MM_Scavenger::workThreadGarbageCollect(this=0x000000012a80dc20, env=0x000000012b827668) at Scavenger.cpp:2611:5 [opt]
    frame #15: 0x0000000100a86408 libj9gc29.dylib`MM_ParallelDispatcher::workerEntryPoint(this=0x0000600000d842c0, env=0x000000012b827668) at ParallelDispatcher.cpp:187:23 [opt]
    frame #16: 0x0000000100a86234 libj9gc29.dylib`dispatcher_thread_proc2(portLib=<unavailable>, info=0x000000016fece460) at ParallelDispatcher.cpp:93:15 [opt]
    frame #17: 0x000000010029559c libj9prt29.dylib`omrsig_protect(portLibrary=0x00000001001d1ed8, fn=(libj9gc29.dylib`dispatcher_thread_proc2(OMRPortLibrary*, void*) at ParallelDispatcher.cpp:58), fn_arg=0x000000016fece460, handler=(libj9vm29.dylib`structuredSignalHandlerVM at gphandle.c:681), handler_arg=0x000000012b015620, flags=506, result=<unavailable>) at omrsignal.c:425:12 [opt]
    frame #18: 0x0000000100a862b4 libj9gc29.dylib`dispatcher_thread_proc(info=<unavailable>) at ParallelDispatcher.cpp:131:2 [opt]
    frame #19: 0x0000000100202644 libj9thr29.dylib`thread_wrapper(arg=0x0000000129037660) at omrthread.c:1733:2 [opt]
    frame #20: 0x000000019d0282e4 libsystem_pthread.dylib`_pthread_start + 136
```

**Comment by thallium** [comment]
I played around with test `java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java`.  I reduced the `num` from 100000 to 10 so there will only be 11 vthreads but I'm still getting an intermittent hang: some threads were stuck in the YIELDING state
after calling `Object.wait()` and were failed to be notified.

```
> !j9object 0x00000003007A7030
!J9Object 0x00000003007A7030 {
        struct J9Class* clazz = !j9class 0x139E27600 // java/lang/VirtualThread
        Object flags = 0x00000008;
        J lockword = 0x0000000000000000 (offset = 0) (java/lang/Object) <hidden>
        J eetop = 0x0000000000000000 (offset = 8) (java/lang/Thread)
        J tid = 0x0000000000000017 (offset = 16) (java/lang/Thread)
        Ljava/lang/String; name = !fj9object 0x300018f68 (offset = 48) (java/lang/Thread)
        Z interrupted = 0x00000000 (offset = 152) (java/lang/Thread)
        Ljava/lang/ClassLoader; contextClassLoader = !fj9object 0x300030ec0 (offset = 56) (java/lang/Thread)
        Ljava/lang/Thread$FieldHolder; holder = !fj9object 0x0 (offset = 64) (java/lang/Thread)
        Ljava/lang/ThreadLocal$ThreadLocalMap; threadLocals = !fj9object 0x0 (offset = 72) (java/lang/Thread)
        Ljava/lang/ThreadLocal$ThreadLocalMap; inheritableThreadLocals = !fj9object 0x0 (offset = 80) (java/lang/Thread)
        Ljava/lang/Object; scopedValueBindings = !fj9object 0x30000b7e8 (offset = 88) (java/lang/Thread)
        Ljava/lang/Object; interruptLock = !fj9object 0x3007a7140 (offset = 96) (java/lang/Thread)
        Ljava/lang/Object; parkBlocker = !fj9object 0x0 (offset = 104) (java/lang/Thread)
        Lsun/nio/ch/Interruptible; nioBlocker = !fj9object 0x0 (offset = 112) (java/lang/Thread)
        Ljdk/internal/vm/Continuation; cont = !fj9object 0x0 (offset = 120) (java/lang/Thread)
        Ljava/lang/Thread$UncaughtExceptionHandler; uncaughtExceptionHandler = !fj9object 0x0 (offset = 128) (java/lang/Thread)
        J threadLocalRandomSeed = 0x0000000000000000 (offset = 24) (java/lang/Thread)
        I threadLocalRandomProbe = 0x00000000 (offset = 156) (java/lang/Thread)
        I threadLocalRandomSecondarySeed = 0x00000000 (offset = 160) (java/lang/Thread)
        Ljdk/internal/vm/ThreadContainer; container = !fj9object 0x3006f9760 (offset = 136) (java/lang/Thread)
        Ljdk/internal/vm/StackableScope; headStackableScopes = !fj9object 0x0 (offset = 144) (java/lang/Thread)
        Z started = 0x00000000 (offset = 164) (java/lang/Thread)
        J tls = 0x0000000000000000 (offset = 40) (java/lang/Thread) <hidden>
        J internalSuspendState = 0x0000000000000000 (offset = 32) (java/lang/Thread) <hidden>
        Ljava/util/concurrent/Executor; scheduler = !fj9object 0x3006f98b0 (offset = 184) (java/lang/VirtualThread)
        Ljdk/internal/vm/Continuation; cont = !fj9object 0x3007a7150 (offset = 192) (java/lang/VirtualThread)
        Ljava/lang/Runnable; runContinuation = !fj9object 0x3007a71d0 (offset = 200) (java/lang/VirtualThread)
        I state = 0x00000010 (offset = 240) (java/lang/VirtualThread)
        Z parkPermit = 0x00000000 (offset = 244) (java/lang/VirtualThread)
        Z blockPermit = 0x00000000 (offset = 248) (java/lang/VirtualThread)
        Z onWaitingList = 0x00000000 (offset = 252) (java/lang/VirtualThread)
        Ljava/lang/VirtualThread; next = !fj9object 0x0 (offset = 208) (java/lang/VirtualThread)
        Z notified = 0x00000000 (offset = 256) (java/lang/VirtualThread)
        B timedWaitSeqNo = 0x00000000 (offset = 260) (java/lang/VirtualThread)
        J timeout = 0x0000000000000000 (offset = 168) (java/lang/VirtualThread)
        Ljava/util/concurrent/Future; timeoutTask = !fj9object 0x0 (offset = 216) (java/lang/VirtualThread)
        Ljava/lang/Thread; carrierThread = !fj9object 0x0 (offset = 224) (java/lang/VirtualThread)
        Ljava/util/concurrent/CountDownLatch; termination = !fj9object 0x0 (offset = 232) (java/lang/VirtualThread)
        J inspectorCount = 0x0000000000000000 (offset = 176) (java/lang/VirtualThread) <hidden>
}
```

**Comment by babsingh** [comment]
https://github.com/eclipse-openj9/openj9/pull/21494 was merged yesterday. See if it helps with the missing notification. 

Also, cherry-pick https://github.com/eclipse-openj9/openj9/pull/21459 into your local workspace.

**Comment by thallium** [comment]
Unfortunately they didn't help.

**Comment by babsingh** [comment]
@fengxue-IS There is another issue with `notify` which didn't get fixed by https://github.com/eclipse-openj9/openj9/pull/21494.

**Comment by fengxue-IS** [comment]
@thallium after reducing the number of thread, can you add `J9VMInternals.dumpString("") to `VirtualThread.afterYield()` call to verify the state of each virtual thread one it is unmounted. then add printf to `BytecodeInterpreter::inlObjectNotify` / `VM_ContinuationHelpers::notifyVirtualThread` to log all the virtualthread that have been notified. lastly add printf to enterContinuation when it is actually remounted, this helps to identify where the notify is failing (ie failed to notify, failure to unblock, failed to re schedule)

**Comment by thallium** [comment]
I'm seeing the owned [monitor count assertion](https://github.com/eclipse-openj9/openj9/blob/master/runtime/vm/ContinuationHelpers.cpp#L1091) added in #21533 being triggered constantly. @tajila FYI.

**Comment by keithc-ca** [comment]
Perhaps #21527 is relevant?

**Comment by thallium** [comment]
I'm using the latest changes.

**Comment by thallium** [comment]
With https://github.com/eclipse-openj9/openj9/pull/21569 the test `java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java` can finish with num=10, but with num=100 it will still hang intermittently.

**Comment by thallium** [comment]
In the test `java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java` sometimes all the threads are hanging in `Object.waitImpl`. I'm seeing the fast JNI path being called which might prevent vthreads from yielding. @fengxue-IS FYI.

**Comment by fengxue-IS** [comment]
@0xdaryl Can you confirm if fastJNI optimization for `Object.wait` have been disabled for virtual threads in JDK24+?

**Comment by tajila** [comment]
> @0xdaryl Can you confirm if fastJNI optimization for Object.wait have been disabled for virtual threads in JDK24+?

It should have been disable with https://github.com/eclipse-openj9/openj9/pull/21489.

If its still hitting the fastJNI path then maybe @gacholio has some suggestions

**Comment by thallium** [comment]
`java/lang/Thread/virtual/stress/LotsOfContendedMonitorEnter.java` is passing with `-Xint`. 20x grinder: https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/49751/

**Comment by fengxue-IS** [comment]
@thallium can you try to update `FastJNI_java_lang_Object.cpp` and ifdef the two following code block with `#if JAVA_SPEC_VERSION < 24`
```C++
void JNICALL
Fast_java_lang_Object_wait(J9VMThread *currentThread, j9object_t receiverObject, jlong millis, jint nanos)
{
...
}

``` 
```c++
J9_FAST_JNI_METHOD("waitImpl", "(JI)V", Fast_java_lang_Object_wait,
		J9_FAST_JNI_RETAIN_VM_ACCESS | J9_FAST_JNI_DO_NOT_WRAP_OBJECTS)
```

and check if the fast jni path is still being used during the test case.

**Comment by thallium** [comment]
After disabling FastJNI path for object wait using the above changes by Jack, the test `java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java` will either hang very intermittently or produce Invalid JIT return address error. @babsingh @tajila do you have any insight for the Invalid JIT return address error?

**Comment by babsingh** [comment]
> do you have any insight for the Invalid JIT return address error?

- Start by looking at the `!stackslots` of the thread, where the crash occurred.
- Inspect the frames. 
- If a virtual thread is involved, check for corruption in the JIT artifacts stored in `J9VMContinuation`. They can also reside in the corresponding `J9VMThread`.

![Image](https://github.com/user-attachments/assets/dec909e3-5b01-43e4-ada1-d3d624119dc8)

- Sometimes, it is related to OSR ... check for tracepoints related to decompilations using the `!snapformat` cmd.

We can involve the JIT team once sufficient evidence is gathered.

**Comment by gacholio** [comment]
Two courses of action here:

1) Break up the big if into nested single ifs and put in prints to see where it fails
2) Use the debugger (much more work IMO)

**Comment by thallium** [comment]
The test `java/lang/Thread/virtual/stress/Skynet100kWithMonitors.java` is passing ([20x grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/50008/)) with the changes https://github.com/eclipse-openj9/openj9/pull/21753 and with `-Xint`. @tajila do you think we should get the changes merged?

**Comment by tajila** [comment]
> do you think we should get the changes merged?

Does the test have an -Xint only variation?

**Comment by thallium** [comment]
> Does the test have an -Xint only variation?

No.

**Comment by tajila** [comment]
@thallium which platform were you seeing the failure?

**Comment by thallium** [comment]
@tajila It's failing for me on both amac and xlinux, here's a [grinder](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/50152/) where it fails

**Comment by tajila** [comment]
You should verify with a grinder, I only did a few runs. Try xlinux first, then other platforms.

**Comment by JasonFengJ9** [comment]
[openjdk24_j9_sanity.openjdk_ppc64_aix](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk24_j9_sanity.openjdk_ppc64_aix_Personal_testList_0/13/consoleFull)
```
09:24:02  variation: -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage Mode150
09:24:02  JVM_OPTIONS:  -Xdump:system:none -Xdump:heap:none -Xdump:system:events=gpf+abort+traceassert+corruptcache -XX:-JITServerTechPreviewMessage -XX:+UseCompressedOops -Xverbosegclog 

10:20:21  TEST: java/lang/Thread/virtual/stress/LotsOfContendedMonitorEnter.java#LM_LIGHTWEIGHT

10:20:21  14:20:10.250 0x302ad800    j9vm.227    *   ** ASSERTION FAILED ** at /home/jenkins/workspace/Build_JDK24_ppc64_aix_Personal/openj9/runtime/vm/classsupport.c:1295: ((vmThread)->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS)

10:33:07  jdk_lang_0_FAILED
```

**Comment by pshipton** [comment]
Also
https://openj9-jenkins.osuosl.org/job/Test_openjdk24_j9_sanity.openjdk_ppc64_aix_Nightly_testList_0/73/ 
https://openj9-jenkins.osuosl.org/job/Test_openjdk24_j9_sanity.openjdk_ppc64le_linux_Nightly_testList_0/73/
https://openj9-jenkins.osuosl.org/job/Test_openjdk24_j9_sanity.openjdk_x86-64_mac_Nightly_testList_1/71/
and other child jobs on these platforms

@thallium pls re-exclude the tests on AIX, plinux, xmac,
